src/nvim/api/buffer.c

```diff
@@ -738,6 +742,219 @@ ArrayOf(Integer, 2) nvim_buf_get_mark(Buffer buffer, String name, Error *err)
+/// @param namespace An id returned previously from nvim_create_namespace
+/// @param lower One of: extended mark id, (row, col) or -1
+/// @param upper One of: extended mark id, (row, col) or -1
+/// @param amount Maximum number of extmarks to return or -1
+/// @param reverse Decides the search direction in the range
+/// @param[out] err Details of an error that may have occurred
+/// @return [[mark_id, row, col], ...]
+ArrayOf(Object) nvim_buf_get_marks(Buffer buffer,
+                                   Integer namespace,
+                                   Object lower,
+                                   Object upper,
+                                   Integer amount,
+                                   Boolean reverse,
+                                   Error *err)
+    FUNC_API_SINCE(1)
+{
+  Array rv = ARRAY_DICT_INIT;
+
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+
+  if (!buf) {
+    return rv;
+  }
+
+  if (!ns_initialized((uint64_t)namespace)) {
+    api_set_error(err, kErrorTypeValidation, _("Invalid mark namespace"));
+    return rv;
+  }
+
+  if (amount == 0) {
+    return rv;
+  }
+
+  linenr_T l_lnum;
+  colnr_T l_col;
+  if (!set_extmark_index_from_obj(buffer, namespace, lower, &l_lnum, &l_col,
+                                  err)) {
+    return rv;
+  }
+
+  linenr_T u_lnum;
+  colnr_T u_col;
+  if (!set_extmark_index_from_obj(buffer, namespace, upper, &u_lnum, &u_col,
+                                  err)) {
+    return rv;
+  }
+
+  // TODO: assert lower <= upper
+
+  ExtendedMark *extmark;
+  Array mark = ARRAY_DICT_INIT;
+
+  // Range Query
+  ExtmarkArray extmarks_in_range;
+
+  extmarks_in_range = extmark_get(buf,
+                                  (uint64_t)namespace,
+                                  l_lnum,
+                                  l_col,
+                                  u_lnum,
+                                  u_col,
+                                  (int64_t)amount,
+                                  reverse ? BACKWARD: FORWARD);
+
+  size_t n = kv_size(extmarks_in_range);
+  for (size_t i = 0; i < n; i++) {
+    mark.size = 0;
+    mark.capacity = 0;
+    mark.items = 0;
+    extmark = kv_A(extmarks_in_range, i);
+    ADD(mark, INTEGER_OBJ((Integer)extmark->mark_id));
+    ADD(mark, INTEGER_OBJ(extmark->line->lnum));
+    ADD(mark, INTEGER_OBJ(extmark->col));
+    ADD(rv, ARRAY_OBJ(mark));
+  }
+  kv_destroy(extmarks_in_range);
+  return rv;
+
+}
+
+/// Create or update a mark at a position
+///
+/// @param buffer The buffer handle
+/// @param namespace a identifier returned previously with extmark_ns_create
+/// @param id The mark's id or ""
+/// @param row position of the mark
+/// @param col position of the mark
+/// @param[out] err Details of an error that may have occurred
+/// @return 1 on new, 2 on update; or a mark_id if argument mark_id was ""
+Integer nvim_buf_set_mark(Buffer buffer,
+                          Integer namespace,
+                          Object mark_id,
```

Comment:
old comment lost in rebase: just use Integer and make `make_id=0` mean "no mark". In general 1 is the first of a handle and 0 is sentinel value.

Reply:
done